homeassistant/components/drop_connect/coordinator.py

```diff
@@ -0,0 +1,66 @@
+"""DROP device data update coordinator object."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import logging
+from typing import TYPE_CHECKING
+
+import attr
+from dropmqttapi.mqttapi import DropAPI
+
+from homeassistant.components import mqtt
+from homeassistant.components.mqtt import ReceiveMessage
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import CONF_DATA_TOPIC, CONF_DEVICE_TYPE, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DROPDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """DROP device object."""
+
+    _drop_api: DropAPI | None = None
+    _unsubscribe_callback: Callable[[], None] | None = attr.ib()
+
+    def __init__(self, hass: HomeAssistant, unique_id: str) -> None:
+        """Initialize the device."""
+        super().__init__(hass, _LOGGER, name=f"{DOMAIN}-{unique_id}")
+        if TYPE_CHECKING:
+            assert self.config_entry is not None
+        self._drop_api = DropAPI()
+        hass.async_create_task(self.subscribe_mqtt(hass))
```

Comment:
I _think_ this is counted as a side effect. I know for sure that we don't allow that. I think it would be better to call this from outside of the coordinator

Reply:
@joostlek Would it be appropriate to subscribe in `async_setup_entry()` in `__init__.py`?

Reply:
I think that works yes!

Reply:
You could use a dispatcher and a listener to the coordinator to process the updates.

An example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/dsmr/sensor.py#L560-L562

Reply:
This is only to subscribe, not really to process update

Reply:
> @joostlek Would it be appropriate to subscribe in `async_setup_entry()` in `__init__.py`?
This works:

```python
async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
    """Set up DROP from a config entry."""

    # Make sure MQTT integration is enabled and the client is available.
    if not await mqtt.async_wait_for_mqtt_client(hass):
        _LOGGER.error("MQTT integration is not available")
        return False

    if TYPE_CHECKING:
        assert config_entry.unique_id is not None
    drop_data_coordinator = DROPDeviceDataUpdateCoordinator(
        hass, config_entry.unique_id
    )

    @callback
    def mqtt_callback(msg: ReceiveMessage) -> None:
        """Pass MQTT payload to DROP API parser."""
        if drop_data_coordinator.drop_api.parse_drop_message(
            msg.topic, msg.payload, msg.qos, msg.retain
        ):
            drop_data_coordinator.async_set_updated_data(None)

    config_entry.async_on_unload(
        await mqtt.async_subscribe(
            hass, config_entry.data[CONF_DATA_TOPIC], mqtt_callback, 0
        )
    )
    _LOGGER.debug(
        "Entry %s (%s) subscribed to %s",
        config_entry.unique_id,
        config_entry.data[CONF_DEVICE_TYPE],
        config_entry.data[CONF_DATA_TOPIC],
    )

    hass.data.setdefault(DOMAIN, {})[config_entry.entry_id] = drop_data_coordinator
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
    return True
```

Reply:
@jbouwh Before I saw your last comment, I committed changes in 414b4dbe5599a4135dc0f54b831d09493c0368d4 that leaves the subscription in a function within the coordinator, but calls it from async_setup_entry().  Please take a look at the current state and let me know if it's acceptable, or if I need to move the contents of subscribe_mqtt() to where it's called in async_setup_entry().

Reply:
It is not directly an issue but it makes it less transparant and now you need 2 async calls, instead of 1. Futher you should assign self.drop_api is stead of implementing it as a property.

Reply:
Moved all of the subscription logic out of the coordinator in 43e8922882d849a26f99d3de9ff77e5b40fcbd8c.  Still working on the other aspects.

Reply:
599e58c2197c2d54bbd190000dfea160ec57e87e takes care of assigning self.drop_api.